### ISSUE TITLE
App crashes on ServerMiddleware hot reload if Mongoose models are used, gives an error like "ServerMiddleware Error: Cannot overwrite User model once compiled"

### ISSUE THREAD

**Issue by kamran-12** [nuxt/nuxt.js#7917]
### Versions

- nuxt: v2.14.1
- node: v12.18.3

### Reproduction
Use [this project](https://github.com/nuxt/nuxt.js/files/5074173/reproduction.zip) to reproduce, you may need to add a working MongoDB database credentials in .env file.

### Steps to reproduce
npm install
npm run dev
(No error at this point yet)
Make a change in login.js file
Save

### What is Expected?
Normal hot reload of servermiddleware as always (like when Mongoose is not used).

### What is actually happening?
App crashes with this error:

`ERROR ServerMiddleware Error: Cannot overwrite User model once compiled.`

**Comment by farnabaz** [comment]
This isn't related to Nuxt, This is mongoose error. When a change occurs and Nuxt reload model files and mongoose models call twice.

Here you can find a fix: https://stackoverflow.com/questions/19051041/cannot-overwrite-model-once-compiled-mongoose

**Comment by kamran-12** [comment]
@farnabaz This is actually a problem with how reloads are implemented with nuxt. I looked at the stackoverflow example you gave me before starting to write this bug report, and although error is same, reasons for it are different.

In stackoverflow example: person with name Anathema.Imbued made a mistake by defining the same model twice within the project, got error immediately (not just on reload).

In my example: didn't do such mistake, same mongoose code works well outside nuxt, also works well inside nuxt **until reload**. Crashes only during hot reload of server middleware.

Please re-open.

**Comment by farnabaz** [comment]
Nodemon restarts the whole server on file change, but Nuxt removes the cache and re-import the file. I believe this is not a bug its nuxt approach to reload files.  
Restarting the server on every file change has a great impact on performance.  

You can avoid mongoose error by checking models to see if `User` model was registered or not.
```js
let modelName = ...
let schema = ...
const Model = mongoose.models[modelName] || mongoose.model(modelName, schema);

export default Model;
```

**Comment by pi0** [comment]
This issue is both with nuxt HMR and mongoose. Using default connection, it keeps instance in CJS cache and when reloading app, it causes. 

> You can avoid mongoose error by checking models to see if User model was registered or not.

@farnabaz I guess this will cause not applying schema changes.

One workaround is doing full reload for server by directly importing serveriMiddleware like this:

```js
export default {
  serverMiddleware: [
    require('./api')
  ]
}
```

Maybe we can also introduce an option to HMR some node_module dependencies. Currently we don't clear cache for all `node_modules` ([src](https://github.com/nuxt/nuxt.js/blob/dev/packages/utils/src/cjs.js#L8) which is also not configurable but if it is, we can specify mongoose or other HMR incompatible dependencies.

**Comment by farnabaz** [comment]
> @farnabaz I guess this will cause not applying schema changes.

Yes, you are right.  


> we can specify mongoose or other HMR incompatible dependencies.

Do you mean to hardcode mongoose and other deps?

**Comment by kamran-12** [comment]
> This issue is both with nuxt HMR and mongoose. Using default connection, it keeps instance in CJS cache and when reloading app, it causes this error
> 
> > You can avoid mongoose error by checking models to see if User model was registered or not.
> 
> @farnabaz I guess this will cause not applying schema changes.
> 
> One workaround is doing full reload for server by directly importing serveriMiddleware like this:
> 
> ```js
> export default {
>   serverMiddleware: [
>     require('./api')
>   ]
> }
> ```
> 
> Maybe we can also introduce an option to HMR some node_module dependencies. Currently we don't clear cache for all `node_modules` ([src](https://github.com/nuxt/nuxt.js/blob/dev/packages/utils/src/cjs.js#L8)) which is also not configurable but if it is, we can specify mongoose or other HMR incompatible dependencies.

I think what you suggested can work as a good solution for my specific case (if what you suggested is full reload of servermiddleware only, not the whole nuxt server), as nodemon restart takes ~1 second while full nuxt reload takes ~5 to 10 seconds on my computer. Two questions:
1. What does HMR mean? (cannot find in internet)
2. Where the code you suggested should be written?

**Comment by kamran-12** [comment]
@pi0 what can be a temporary solution for me? So that when something inside servermiddleware changes, all of the servermiddleware (but not the whole nuxt server) restarts (like nodemon)? That would be good enough for me (as nodemon restart takes ~1 sec., while full nuxt restart takes 5 to 10 sec (that needs improvement, too, btw)). I guess what you suggested was supposed to be what I am asking for right now, but it does not work (the same error is still there).

**Comment by pi0** [comment]
> Do you mean to hardcode mongoose and other dependencies?

@farnabaz That could be a quick nice patch by modifying regex 

> What can be a temporary solution for me?

@kamran-12 Using `require` as mentioned above, will cause full restart. HMR is to only replace some `requires` on runtime to avoid full restart. Alternatively before fix for mongoose, you can use [concurrently](https://www.npmjs.com/package/concurrently) to run API standalone and make requests from UI to it. (can provide more detailed info how to do)

**Comment by kamran-12** [comment]
@pi0 if I understood right, you suggested to replace `serverMiddleware: [{path: '/api', handler: '~/api/index'}]` to `serverMiddleware: [require('./api')]` in nuxt.config.js. It didn't work. The error didn't go away. I have said this in the last sentence of my previous comment.

> Alternatively before fix for mongoose, you can use concurrently to run API standalone and make requests from UI to it. (can provide more detailed info how to do)

That would be helpful.

**Comment by pi0** [comment]
fixed by [v2.14.4](https://github.com/nuxt/nuxt.js/releases/tag/v2.14.4)

**Comment by liho00** [comment]
Same issue on nuxt@2.15.3 
![image](https://user-images.githubusercontent.com/77238199/117630841-2ea62b80-b1ae-11eb-9e38-2923d314d920.png)

**Comment by danielroe** [comment]
@liho00 Would you open a new issue with a reproduction? 🙏

**Comment by AregGhazaryan** [comment]
Why this has been closed? It still persists in nuxt 2.15.7

**Comment by danielroe** [comment]
@AregGhazaryan If it's still persisting for you, would you open a new issue with a reproduction? 🙏

**Comment by AregGhazaryan** [comment]
@danielroe #9456
